### DISCUSSION TITLE
Replacing TRRS cable with common phone wall charger USB-A to USB-C and making it disaster/fool proof

### DISCUSSION THREAD

**Discussion by kaide** [JellyTitan/Sofle-Pico#13] [Q&A] [answered]
Hi Ryan, thanks for your effort in this project.

I'm preparing your sofle-pico build, and quit paranoid about that TRRS shorting issue especially with kids around. Since I have a lot of spare phone charger cables (USB A to USB C), and there are breakout PCB board available online, I plan to connect the TRRS connection from the Sofle-Pico PCB to the USB breakout PCB.

To prevent disaster when people are connecting their normal USB-C host (from laptop directly to the USB-C TRRS replacement) I plan to not using USB (A or C) VCC pin and GND pin at all, so when user connect their laptop to the USB-C TRRS replacement, nothing will happen.

Is this ok?

Thanks

 
![Female USB-A and USB-C breakout board](https://github.com/JellyTitan/Sofle-Pico/assets/1152826/ef8c0865-6e1d-4b14-aca8-c2a7332376bc)

**Comment by JellyTitan** [comment]
It sounds like you plan to replace the TRRS jacks with a USC-C breakout board, and not connect the default VCC & GND pins that USB-C would normally use, so that if the breakout board is accidentally connected directly to a power source, it will not short? 

Am I understanding that correctly?

Usually USB-C cables between split keyboards are avoided precisely because of the problem you mentioned - there's the potential to connect directly to the power and short the board. 

I've seen some boards that will use USB-C between two halves, like the [Keeb.io iris](https://keeb.io/products/iris-keyboard-split-ergonomic-keyboard). Those boards don't use MCU boards like the Pico or the ProMicro, those circuits are all on one PCB, which make me think there is some voltage protection between the two halves.  

I think a better solution would be a RJ45 jack, or you can wire the halves together directly.

**Reply by kaide** [reply]
Yes that is what I want to prevent: People using the phone wall charger or laptop to connect to the USB-C TRRS replacement and short the board.

In theory if I do not use the USB VCC and GND; when people connecting the wall charger to the USB-C TRRS replacement, shouldn't the current not flowing from the charger? thus preventing the overvoltage/short?

I also looked ad RJ45 jack, but the available PCB breakout is not as small as the USB one :)

**Reply by JellyTitan** [reply, marked as answer]
Your concern is valid, but I'm not sure using a modified USB config is the best option. @TheWerle has a great idea for TRRS backpower protection. I've converted that into an issue:
https://github.com/JellyTitan/Sofle-Pico/issues/15

**Comment by TheWerle** [comment]
[I'm actively looking to develop updated/improved best practices for this issue](https://www.reddit.com/r/ErgoMechKeyboards/comments/1aupzz3/split_keyboard_best_practices_for_protection/), but here's my current two-cents.

For reducing the dual power source risk I definitely endorse either of the Zener-diode or PFET protection solutions in RP2040 datasheet section 4.5 (pg22).  That is what I'm proposing in [https://github.com/JellyTitan/Sofle-Pico/issues/15] and you could easily wire it into a breakout adapter solution based on the current Pico SCH/PCB.  

TRRS inherently risks shorts due to the sequential mechanical mating of the contacts between the plug and jack. I too want some added safety mechanisms for kids/my own clumsiness. For implementing a "safe break" link I'm going to research various magnetic-jack connectors and cables, but that's a whole other thing.

Without product links and some rough schematics/wiring diagrams for your implementation I can't fully endorse any generic breakout board, it all depends on both how the vendor routes the PCB to their connector, the connector's solder-contact types and board mounting reinforcement, and how you'd then wire the adapter into the keyboard.  Conceptually its fine, but there's potentially devils in the details. 

I personally dislike the USB-C cable solution for the wiring ambiguity it brings and the nebulous improvements for mechanical strength at the connector. That image shows four through-hole mechanical connections to the USB-C housing, which at least implies robustness, but I can't tell how the electrical connections route. If the breakout the full USB-C pinout properly assigned for reversible insertion it should be generally OK, but you effectively trade TRRS electrical shorting risk for dual-power risk when somebody hooks the keyboard up to the wrong USB-C.  That would in turn be mitigated by adding the diode/PFET solution, but requires the design tradeoff of rigid primary/secondary keyboard halves that determine your power-flow. How you personally weigh those tradeoffs is up to you.

**Reply by kaide** [reply]
Yes, now I get it, why it is still possible to have 5v into the USB-C replacement for TRRS even though you disconnect the VCC and GND and use other pin.

Because the other end can be 4 wires cable with 5 Volt.

![image](https://github.com/JellyTitan/Sofle-Pico/assets/1152826/a7a1b75d-9818-4408-b81d-6cd21f79d983)

**Reply by kaide** [reply]
I'll use RJ-45 for now. Will use this breakout board or make one using standard grid holed PCB
![image](https://github.com/JellyTitan/Sofle-Pico/assets/1152826/efe1bae0-629e-402c-a8b6-c3b3f3358e67)

**Reply by Luciogi** [reply]
can you share pin diagram for using rj 45?

**Reply by TheWerle** [reply]
> can you share pin diagram for using rj 45?

It would be an an arbitrary choice, as you're not adhering to a standard.  Lots of keyboards use older RJ-11 four-wire connectors for the same purpose because they're functionally identical but smaller.  RJ-45 is just a bigger, eight-wire equivalent that's more readily available in modern times.

**Reply by Luciogi** [reply]
I don't have knowledge of electronic DIY.

My guess, is RJ 45 will go into this area
<img width="373" height="417" alt="image" src="https://github.com/user-attachments/assets/128e835b-c8a1-4be9-a094-ba3396af8957" />


but not sure how many pin from Rj 45 will go into pcb.

**Reply by Luciogi** [reply]
I can see rj 11 with 6 pins are available in my country.

**Reply by TheWerle** [reply]
Neither RJ-45 and RJ-11 are intended to be used with this design.  Any connector with four wires (VSYS, GND, TX, RX) can be made to work if, but only TRRS is a simple solution.   If you can get a TRRS cable/connectors just install and use them as intended.  For a beginner wiring in an alternative connector is not worth the cost/risk of buying and soldering alternatives. Keep it simple.

**Reply by Luciogi** [reply]
I want to avoid MCU pin burning problem due to yanking of trrs cable

**Comment by Luciogi** [comment]
Here is my first attempt to create breakout board
<img width="1404" height="632" alt="image" src="https://github.com/user-attachments/assets/e75ec5cb-f61f-4734-91bf-d40df06ed159" />
<img width="1529" height="542" alt="image" src="https://github.com/user-attachments/assets/46903eb8-0a01-4f10-a2a5-57ba2dcbe090" />
<img width="1262" height="429" alt="image" src="https://github.com/user-attachments/assets/4a8e58ac-a489-412d-94f8-4cb12644e1f5" />
<img width="1100" height="392" alt="image" src="https://github.com/user-attachments/assets/d96a8c3d-3d77-415e-b445-a46fe94b62f1" />

**Comment by Luciogi** [comment]
Here is 2nd attempt

I did mistake that Both sides were connected like TX - TX and  RX - RX
Now I have fixed that by TX - RX and RX - TX

My RJ 11 / 12 is in compact design, so I found RJ14_Connfly_DS1133-S4_Horizontal, which has exact dimensions as mine.
Previously I was using RJ12_Amphenol_54601
<img width="1139" height="400" alt="image" src="https://github.com/user-attachments/assets/31df7d35-63d5-4521-a57e-6628b6e6b800" />

<img width="1599" height="617" alt="image" src="https://github.com/user-attachments/assets/f1957533-2713-49dc-906a-ea2efeee4660" />


In Sofle Pico PCB, I saw that VSYS and GND tracks have width of `0.5mm`. And currently I am using track width of `0.2mm`.

My Question is do I need to change track width to `0.5mm ` for GND and VSYS ?

**Reply by Luciogi** [reply]
@TheWerle  Can you review this solution?

Do find any issues?

**Reply by TheWerle** [reply]
You don't need to deal with RX/TX inversion in SCH/PCB, QMK "config.h" reassigns those signals between the Split halves so people can design reversible PCBs.  

Wider traces on power nets are good practice.  You don't need 0.5mm, but I'd recommend a minimum of 0.25-0.3mm if able.  Another "best practice" is to not route digital signal lines with 90deg angles.  I don't know if the TX trace splitting the testpoints on the righthand side has sufficient copper-to-copper clearance, for typical prototype PCB fab you want a minimum of 0.125mm. 

Unless you've got a way to homebrew these PCBs I'd just focus on a mechanical solution to fixture the connectors and hand-wire it, TBH. Seems like a fairly costly solution to a 4-wire problem.

**Reply by JellyTitan** [reply]
I agree with theWerle on this one - an extra board might be more trouble then creating a sofle pico variant. There was a board that natively used RJ45, i think it was Cheapino? The selling point for that board was that everything was cheap. I think Cat5 cables are probably cheaper and easier to come by than TRRS nowadays. I think that an RJ45 variant would be neat - but I don't have it in me to go through another round of prototyping myself.
